firewood/src/merkle/proof.rs

```diff
@@ -258,261 +257,146 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
+    fn proof_to_path<K: AsRef<[u8]>, S: ShaleStore<Node> + Send + Sync, T: BinarySerde>(
         &self,
-        key: KV,
+        key: K,
         root_hash: [u8; 32],
         merkle_setup: &mut MerkleSetup<S, T>,
         allow_non_existent_node: bool,
     ) -> Result<Option<Vec<u8>>, ProofError> {
         // Start with the sentinel root
-        let root = merkle_setup.get_root();
+        let sentinel = merkle_setup.get_sentinel();
         let merkle = merkle_setup.get_merkle_mut();
-        let mut u_ref = merkle.get_node(root).map_err(|_| ProofError::NoSuchNode)?;
+        let mut parent_node_ref = merkle
+            .get_node(sentinel)
+            .map_err(|_| ProofError::NoSuchNode)?;
 
-        let mut chunks = Vec::new();
-        chunks.extend(key.as_ref().iter().copied().flat_map(to_nibble_array));
+        let mut key_nibbles = Nibbles::<0>::new(key.as_ref()).into_iter();
 
-        let mut cur_key: &[u8] = &chunks;
-        let mut cur_hash = root_hash;
+        let mut child_hash = root_hash;
         let proofs_map = &self.0;
-        let mut key_index = 0;
-        let mut branch_index = 0;
-
-        loop {
-            let cur_proof = proofs_map
-                .get(&cur_hash)
-                .ok_or(ProofError::ProofNodeMissing)?;
-            // TODO(Hao): (Optimization) If a node is already decode we don't need to decode again.
-            let (mut chd_ptr, sub_proof, size) =
-                self.decode_node(merkle, cur_key, cur_proof.as_ref(), false)?;
+        let mut child_index = 0;
 
+        let sub_proof = loop {
             // Link the child to the parent based on the node type.
-            match &u_ref.inner() {
+            // if a child is already linked, use it instead
+            let child_node = match &parent_node_ref.inner() {
                 #[allow(clippy::indexing_slicing)]
-                NodeType::Branch(n) => match n.chd()[branch_index] {
+                NodeType::Branch(n) => match n.chd()[child_index] {
                     // If the child already resolved, then use the existing node.
-                    Some(node) => {
-                        chd_ptr = node;
-                    }
+                    Some(node) => merkle.get_node(node)?,
                     None => {
+                        let child_node = decode_subproof(merkle, proofs_map, &child_hash)?;
+
                         // insert the leaf to the empty slot
                         #[allow(clippy::unwrap_used)]
-                        u_ref
-                            .write(|u| {
-                                let uu = u.inner_mut().as_branch_mut().unwrap();
+                        parent_node_ref
+                            .write(|node| {
                                 #[allow(clippy::indexing_slicing)]
-                                (uu.chd_mut()[branch_index] = Some(chd_ptr));
+                                let node = node.inner_mut().as_branch_mut().unwrap();
```

Comment:
I can use `expect`. This pattern is used everywhere because it's the only way to write to the `inner` value of the node. We pattern match on `node.inner` but have to call `write` on `node`.